src/ngraph/validation_util.cpp

```diff
@@ -651,221 +651,95 @@ PartialShape ngraph::infer_slice_shape(const Node* node,
+            int num_input_axis_before_ellipses = 0;
+            for (size_t i = 0; i < axis; ++i)
             {
-                begin_dms[i] = input_shape[j] - 1;
+                if (!new_axis_mask.count(i))
+                    num_input_axis_before_ellipses++;
             }
-        }
-
-        // Use upper_bounds if mask is not set
-        if (ub_mask.find(j) == ub_mask.end())
-        {
-            Dimension end_dms_tmp;
-
-            if (ub.size() <= ej)
-            {
-                end_dms_tmp = end_dms[i];
-            }
-            else if (stride_dms[i].is_dynamic())
-            {
-                end_dms_tmp = Dimension::dynamic();
-            }
-            else if (int64_t(stride_dms[i]) > 0)
-            {
-                end_dms_tmp = ub[ej] - 1;
-            }
-            else
-            {
-                end_dms_tmp = ub[ej] + 1;
+            for (size_t i = axis + 1; i < begin.size(); ++i) {
+                if (new_axis_mask.count(i))
+                    num_new_axis_after_ellipses++;
             }
 
-            if (ub.size() > ej)
-            {
-                end_dms[i] = end_dms_tmp;
-            }
-            else if (stride_dms[i].is_dynamic())
-            {
-                end_dms[i] = Dimension::dynamic();
-            }
-            else if (int64_t(stride_dms[i]) > 0)
+            unsigned long num_input_axis_after_ellipses = (begin.size() - axis - num_new_axis_after_ellipses - 1); // -1 because it's a position of ellipses
+            unsigned long num_of_hidden_dims = input_shape.to_shape().size() - num_input_axis_after_ellipses
+                    - num_input_axis_before_ellipses;
+            for (size_t i = 0; i < num_of_hidden_dims; ++i)
             {
-                end_dms[i] = -1;
-            }
-            else
-            {
-                end_dms[i] = 0;
+                dim.emplace_back(input_shape[input_shape_idx]);
+                input_shape_idx++;
             }
         }
-        else
-        {
-            if (stride_dms[i].is_dynamic())
-            {
-                end_dms[i] = Dimension::dynamic();
-            }
-            else if (int64_t(stride_dms[i]) > 0)
-            {
-                end_dms[i] = -1;
+        else {
+            if (new_axis_mask.count(axis)) { // add new single dimension if new_axis_mask is set
+                dim.emplace_back(1);
             }
-            else
-            {
-                end_dms[i] = 0;
+            else if (shrink_axis_mask.count(axis)) { // skip this dimension if shrink_axis_mask is set
+                input_shape_idx++;
             }
-        }
-        ej++;
+            else { // calculating dimension (begin, end, begin_mask, end_mask, stride)
+                int64_t lb = begin[axis];
+                int64_t ub = end[axis];
+
+                // convert negative indexes to positive
+                if (lb < 0)
+                    lb = std::max(int64_t(input_shape[input_shape_idx]) + lb, 0l);
```

Comment:
Not sure I understand why we want the `max(...,0)` here. Could you add a comment?

Reply:
everything is ready